<div style="border:solid Chocolate 2px; padding: 40px">

<b>Илона, приветствую!</b>

Буду рада с тобой познакомиться. Меня зовут Евгения Маракова и я буду ревьюером твоего проекта. Если ты не против, то предлагаю общаться на "ты":) Если удобнее на "вы", то нет проблем, только скажи об этом.

В процессе работы ты встретишь комментарии <font color='green'>зеленого</font>, <font color='gold'>желтого</font> и <font color='red'>красного</font> цветов. Что они значат, смотри ниже:
    <br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Решение на этом шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Решение на этом шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев мой комментарий, не удаляй его, он будет очень полезен в случае повторной проверки работы😊
    
Свой комментарий помечай так, как показано ниже, тогда нам будет легче вести диалог, а мне выполнить повторную проверку проекта.
    <div class="alert alert-info"> <b>Комментарий студента: </b> </div>
    


# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Илона, молодец, отлично выполнила предобработку!
</div>

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# нас интересуют 2 столбца: children и debt. Вопрос стоит в том, влияет ли КОЛИЧЕСТВО детей на задолженность
# еще раз посмотрим на уникальные значения в столбце детей
print(data['children'].unique())
# мы можем подсчитать количество и процент задолженностей по категориям (бездетные, от 1 до 3 - малодетные, многодетные) или по
# таким категориям(бездетные, 1 ребенок, 2 ребенка, 3 ребенка и т. д.)
# для начала разобъем на 3 категории
def categorize_family(quantity):
    if quantity == 0:
        return 'б/д'
    if 1 <= quantity <= 3:
        return 'до 3х'
    return 'м/д'
data['children_category'] = data['children'].apply(categorize_family)
# подсчитаем сколько всего строк по кличеству детей
group_children = data.groupby('children')['debt'].count()
print(group_children)
# создадим сводную таблицу, в которой будет наглядно показано семейная категория, 
# количество детей, количество строк по наличию или отсутствию задолженности
data_count_children = data.groupby(['children_category', 'children', 'debt']).agg({'debt':'count'})
display(data_count_children)
# создадим табличку, в которой будет видно количество людей, которые имели или не имели задолженность и конверсия 
data_grouped = data.groupby('children').agg({'debt': ['count', 'sum', 'mean']})
display(data_grouped)

[1 0 3 2 4 5]
children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64


debt
children_category children debt       
б/д               0        0     13028
                           1      1063
до 3х             1        0      4364
                           1       444
                  2        0      1858
                           1       194
                  3        0       303
                           1        27
м/д               4        0        37
                           1         4
                  5        0         9

debt                
          count   sum      mean
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Cупер, верно сгруппировала данные для анализа!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Рекомендую использовать функцию `display()` вместо `print()` для вывода таблиц в Jupyter Notebook.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Совсем другой вид:)

</div>

**Вывод:** 
1) Среди количества бездетных клиентов только у 7,5% есть задолженность по кредитам.
2) У многодетных клиентов задолженность составляет 9,7% (это самый высокий показатель, но что интересно, у клиентов с 5 детьми вообще нет задолженностей).
3) Клиенты, которые имеют от 1 до 3 детей примерно в 90% случаях не имеют задолженностей по кредитам (если в семье 1 ребенок, то задолженность равна 9,2%, если 2 ребенка - 9,4%, если 3 ребенка - 8,1%).
На основании результатов этого исследования было выяснено, что количество детей практически не влияет на задолженность по возврату кредитов. От общего числа всех клиентов около 8% людей имеют задолженность перед банком.

<div class="alert alert-info"> 
    <b>Комментарий студента: 
    
</b>
В категориях с количеством детей от 3 до 5 представлено наименьшее число клиентов (в сумме их 380 человек, что составляет менее 5% от общего числа). Полученные по ним данные не могут целостно отражать зависимость задолженности по кредитам от количества детей. 
Следует более подробно изучить те категории, в которых выборка клиентов более обширна. К таким категориям относятся бездетные (14091 человек), родители 1 (4808 человек) и 2х детей (2052 человек). Люди, у которых на данный момент времени нет детей, имеют тенденцию к возврату кредита в срок, у них самый меньший показатель задолженности.
С появлением 1 ребенка задолженность составляет уже 9,2%, а у родителей 2х детей процент невозврата в срок выростает до 9,4%
Таким образом, гипотеза о том, что с увеличением числа детей в семье возрастает и задолженность по кредитам может быть подтверждена. Однако следует принять во внимание, что имея категории, несоразмерные по числу клиентов, на данный момент сложно определить эту зависимость. Необходимо проанализировать большее число клиентов с 3-5 детьми.
</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Умничка, верно дополнила!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Стоит обратить внимание на несбалансированность выборок. Поэтому лучше, для выявления зависимости, сравнить самые многочисленные категории - семьи с 0, 1 и 2 детьми. У тебя не хватает общего вывода, если мы все обобщим то, что наблюдаем с ростом количества детей в семьях?
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# ход исследования будет ориентироваться по аналогии с предыдущим заданием
# нас интересуют 2 столбца: family_status и debt. 
print(data['family_status'].unique())
# можно выделить 5 категорий: 'женат / замужем', 'гражданский брак', 'вдовец / вдова', 'в разводе', 'Не женат / не замужем'
# сгруппируем количество клиентов по каждой категории
group_family = data.groupby('family_status')['debt'].count()
print(group_family)
# создадим сводную таблицу, в которой будет наглядно показано семейное положение и факт наличия или отсутствия задолженности
data_count_family = data.groupby(['family_status', 'debt']).agg({'debt':'count'})
display(data_count_family)
# создадим табличку, в которой будет видно количество людей, которые имели или не имели задолженность и конверсия 
data_grouped_f_status = data.groupby('family_status').agg({'debt': ['count', 'sum', 'mean']})
display(data_grouped_f_status)

['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']
family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: debt, dtype: int64


debt
family_status         debt       
Не женат / не замужем 0      2523
                      1       273
в разводе             0      1105
                      1        84
вдовец / вдова        0       888
                      1        63
гражданский брак      0      3749
                      1       385
женат / замужем       0     11334
                      1       927

debt               
                       count  sum      mean
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

**Вывод:**
1) От общего количества клиентов особо выделяются 2 категории: вдовцы и женатые. Для женатых характерно то, что в этой группе наибольшее количество людей (12261), параллельно с этим можно выделить то, что среди них у 7,5% есть задолженность. Что касается овдовевших потребителей - их наименьшее количество среди всех участников выбоки (951). При этом только 6,6% имеют задолженность. Это самый низкий показатель долга по семейному статусу.
2) Доля незамужних клиентов-должников составлет 9,7% (самый высокий показатель из всех), доля разведенных клиентов с показателем задолженности 1 составляет 7,0%, а доля людей в гражданском браке - 9,3%.
Согласно результатам, наименьший показалель задолженности наблюдается у вдовцов/вдов, а самый высокий - у незамужних клиентов. Разница между ними составляет 3,3%. В основном, семейное положение не влияет на задолженность. От общего числа всех клиентов около 8% людей имеют задолженность перед банком.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Но все же, можем сказать, что заемщики состоящие или состоявшие в браке реже допускают просрочку!

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# нас интересуют 2 столбца: total_income и debt. Поскольку в первой части работы мы разбили доходы клиентов по категориям, то 
# сейчас будем также пользоваться столбцом total_income_category, где определяется уровень дохода:
# 0–30000 — 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B'; 1000001 и выше — 'A'.
# создадим табличку, в которой будут видны уровень дохода, количество человек в каждй группе,
# количество людей, имеющих задолженность и конверсию
data_count_total = data.groupby('total_income_category').agg({'debt': ['count', 'sum', 'mean']})
display(data_count_total)

debt                
                       count   sum      mean
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

**Вывод:** 
По данным итоговой таблицы можно сделать следующие выводы:
1) Доля клиентов с доходом уровня С имеет средний зароботок и доля задолженности составляет 8,5% от их общего числа в группе (самая большая доля относительно всего числа - 15921 клиент).
2) Следующей по количеству (5014 клиентов) является группа с доходом В (7,0% задолженности)
3) Самые маленькие доли клиентов - люди с уровнем дохода А и Е (у них процент задолженности 8,0% и 9,0% соответсвенно)
4) Процент задолженностей по кредиту у клиентов с уровнем дохода D минимален - 6,0%
Таким образом, если рассматривать все 5 категорий вместе, то уровень дохода почти не влияет на степень задолженности. Также внимание привлекает практически одинаковый процент задолженности у клиентов с доходом А и Е. От общего числа всех клиентов около 8% людей имеют задолженность перед банком.

<div class="alert alert-info"> 
    <b>Комментарий студента: 
    
</b>
Для подтверждения или опровержения гипотезы зависимости уровня дохода и задолженности, более подробно рассмотрим группы клиентов с доходом В и С, поскольку они являются самыми многочисленными и результат их анализа может отличаться от предыдущего вывода. Категория с уровнем дохода от 50.001 до 200.000 - самая большая (15921 человек с процентом задолженности 8,5%), в категории с доходом от 200.001 до 1.000.0000 процент составил 7,0%, но в ней практически в 3 раза меньше клиентов. Даже учитывая это, можно предположить, что с увеличением уровня дохода снижается % людей с кредитным долгом. Для более точного исследования необходимо проанализировать целевые группы, соразмерные по охвату.
    
    
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Согласна, чем выше доход, тем надежнее заемщик!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Илона, стоит обратить внимание на несбалансированность выборок. Сравни, пожалуйста, две самые многочисленные выборки B и C для выявления зависимости.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [31]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# ход исследования будет ориентироваться по аналогии с предыдущим заданием
# нас интересуют 2 столбца: purpose и debt. Ранее мы создали новый столбец, где привели 
# к стандарту все сформулированные по-разному цели клиентов (purpose_category), им мы и будем пользоваться для исследования 
print(data['purpose_category'].unique())
# сгруппируем количество клиентов по каждой из четырех цели
group_purpose = data.groupby('purpose_category')['debt'].count()
print(group_purpose)
# создадим сводную таблицу, в которой будет наглядно показана цель взятия кредита и факт наличия или отсутствия задолженности
data_count_purpose = data.groupby(['purpose_category', 'debt']).agg({'debt':'count'})
display(data_count_purpose)
# создадим табличку, в которой будет видно количество людей, которые имели или не имели задолженность и конверсия 
data_grouped_purpose = data.groupby('purpose_category').agg({'debt': ['count', 'sum', 'mean']})
display(data_grouped_purpose)

['операции с недвижимостью' 'операции с автомобилем'
 'получение образования' 'проведение свадьбы']
purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64


debt
purpose_category         debt      
операции с автомобилем   0     3879
                         1      400
операции с недвижимостью 0     9971
                         1      780
получение образования    0     3619
                         1      369
проведение свадьбы       0     2130
                         1      183

debt               
                          count  sum      mean
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

**Вывод:** 
1) Число людей, цель которых 'операции с недвижимостью', самая многочисленная (в ней 10751 человек) с процентом должников 7,2%.
2) Меньше всего людей интересуется получением финансов на проведение свадьбы (2313 человек), из которых 7,9% имеют долг.
3) На покупку автомобиля и получение образования менее половины запросов от общего числа клиентов. здесь ситуация с фактом задолженности уже другая - 9,3% и 9,2%. 
Исходя из полученных данных можно предположить, что покупка автомобиля или получение образования не оказывают влияние на возврат. От общего числа всех клиентов около 8% людей имеют задолженность перед банком.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Умница, верно определила более и менее рискованные для банка категории заемщиков по целям кредита!
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Пропуски данных были в столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход).
Возможны несколько причин для их появления:
1. Анкеты были неправильно заполнены самими клиентами (забыл стаж; удержание зарплаты и вследствие этого изменение ежемесячного дохода).
2) Анкеты были неправильно заполнены сотрудником банка (забыл сохранить данные, посмотрел не в тот столбец или строку).
3) Возможно, случился какой-то системный сбои и данные не были сохранены или строка в бумажной/электронной форме заполнения не была помечена как обязательная.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Совершенно верно, на пропуски в данных, в основном, оказывают влияние человеческий и технический факторы!
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Если в количественных данных были слишком большие разрывы, то считать среднее арифметическое - не самый удачный вариант, поскольку одно значение может быть слишком маленьким, а другое слишком завышенным. Тогда не будет отражен реальный ход вещей и данные будут усредненными и неправдивыми.
Медиана же показывает промежуточное значение для каждой подгруппы, а это значит что данные не будут противоречить друг другу и отражать действительность. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Если быть точнее, то медиана менее чувствительна к выбросам!
</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.
Для работы были даны данные со статистикой о платёжеспособности клиентов. Для того, чтобы подготовить корректные данные нужно было провести несколько манипуляций с предобработкой:
1) В ходе работы мы воспользовались методом info() и обнаружили, что некоторые данные в таблице пропущены (в столбцах days_employed и total_income), и поскольку ежемесячный доход у разных групп населения и профессий может отличаться, то было принято решение заполнить пропуски медианным доходом среди всех записей с тем же типом. Для этого необходимо было воспользоваться поиском комбинаций строк с уникальным значением занятости и строк с пропущенными значениями. Вместо них перезаписывались медианные значения. 
2) Также в столбце days_employed наблюдались отрицательные значения. Эти значения заменялись положительными с помощью метода abs(), он преобразует число в модуль этого числа.
3) В столбце children наблюдались аномальные значения (количество детей не может быть -1 и вероятность иметь 20 детей тоже чрезвычайно мала), поэтому эти строки было решено убрать вовсе индексацией конкретных строк и методом drop().
4) В столбце education требовалось с помощью метода unique() найти всевозможные написания уровня образования клиента, а с помощью метода str.lower() привели все значения к нижнему регистру, поскольку все данные были записаны строчными и большими буквами.
5) В массиве данных с помощью комбинации методов duplicated() и sum() были найдены продублированные строки и впоследствии удалены методом drop_duplicates().
6) Далее нам потребовалось создание цикла, в котором значения столбца total_income преобразовавалить в категории дохода клиента (A, B, C, D, E), а затем с помощью apply() мы добавили новый столбец с категориями к датафрейму.
7) В столбце purpose методом unique()выявили, что цели взятия кредита были написаны по-разному. Их нужно было также привести к стандартному виду с помощью цикла и метода apply().
По итогам исследования поставленно несколько гипотез, которые нужно было проверить посредством рассчетов. В ходе данного исследования было выяснено: 
1) что задолженность клиентов по каждому из заданных критериев не превышает 10%;
2) при проведении рассчетов следует обращать внимание не только на процент долга к общему числу клиентов, но и относительно каждой группы по отдельности;
3) семейный статус, количество детей, уровень дохода и цель кредита влияют на кредитную задолженность, но крайне незначительно. Требуются более подробные исследования, не только по каждому критерию, но и в комбинации, поскольку это может повлиять на построение модели кредитного скоринга;
4) не стоит забывать о найденных у двух типов занятости (безработные и пенсионеры) аномально больших значений.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Здорово, что собрала все данные в общий вывод!
</div>

<div style="border:solid Chocolate 2px; padding: 40px">
Илона, давай подведем итоги. Ты большая молодец, что старалась и проделала такую большую работу. Шаг 1 и 2 ты выполнила без проблем, что очень хорошо:)
    
    Разберем над чем стоит поработать:
    
    1. Рекомендую:
    - использовать функцию `display()` для вывода таблиц;
    - учитывать несбалансированность выборок.
    
    Пересмотри работу и исправь ошибки. Обязательно пиши, если я могу тебе еще чем-то помочь. Удачи!)
    
<div class="alert alert-info"> 
    <b>Комментарий студента: 
    
</b>
Евгения, спасибо за подробную обратную связь! Исправила вывод таблиц на экран (стало визуально удобнее и проще для понимания) и подробнее описала результаты, учитывая численность категорий для выборки.
    
</div>
  

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Илона, привет! Молодец, оперативно и верно доработала проект. Благодарю за обратную связь, работа принята, удачи в обучении!)

</div>